compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1738,6 +1740,7 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
         // label pointing out the cause for the type coercion will be wrong
         // as prior return coercions would not be relevant (#57664).
         let fn_decl = if let (Some(expr), Some(blk_id)) = (expression, blk_id) {
+            self.explain_self_literal(fcx, &mut err, expr, expected, found);
```

Comment:
I feel like a better location for this is with the rest of the mismatch suggestions in `emit_type_mismatch_suggestions`.